lib/nginx.js

```diff
@@ -0,0 +1,130 @@
+var nginxConf = require('./nginx-conf').genNginxConf;
+var configHash = require('./nginx-conf').getConfigHash;
+var path = require('path');
+var debug = require('debug')('strong-nginx-controller:nginx');
+var childProcess = require('child_process');
+var async = require('async');
+
+function Nginx(options) {
+  this._app = options.app;
+  this._nginxPath = options.nginxPath;
+  this._baseDir = options.baseDir;
+  this._nginxConf = nginxConf.bind(null, options);
+  this._configHash = configHash.bind(null, options);
+}
+
+/**
+ * Generate the nginx config and start nginx.
+ */
+function start(callback) {
+  async.series([
+    this._nginxConf,
+    this._cmd.bind(this, 'start')
+  ], callback);
+}
+Nginx.prototype.start = start;
+
+var MAX_DELAY = 1000 * 1000; // 1sec
+var reloadTimeoutHandle;
+var delayStartTime;
+/**
+ * The reload function can be called multiple times and in parallel if updates
+ * arrive from Central. This function de-bounces the reload calls in order to
+ * prevent the nginx config from being corrupted by multiple read/write and
+ * prevent multiple nginx restarts.
+ */
+function reload(callback) {
+  var self = this;
+
+  if (reloadTimeoutHandle) {
+    if ((+Date.now() - delayStartTime) > MAX_DELAY) {
+      return setImmediate(callback);
+    }
+
+    debug('Debouncing reload');
+    clearTimeout(reloadTimeoutHandle);
+  } else {
+    debug('Nginx reload requested');
+    delayStartTime = +Date.now();
+  }
+
+  reloadTimeoutHandle = setTimeout(function() {
+    debug('Reloading nginx');
+    reloadTimeoutHandle = null;
+    self._reload(function(err) {
+      if (err)
+        console.error('Error reloading Nginx configuration: %s', err.message);
+    });
+  }, 500);
+  setImmediate(callback);
+}
+Nginx.prototype.reload = reload;
```

Comment:
Lodash isn't already a (direct) dependency, but it might be worth adding (hoisting) for [_.debounce](https://lodash.com/docs#debounce).
```js
Nginx.prototype.reload = require('lodash').debounce(reload, 500, {maxWait: 1000});
```

Or maybe [_.throttle](https://lodash.com/docs#throttle) would be more appropriate here?